Execute inserts inside ExecutorService

mgazanayi opened the pull request.

mgazanayi committed:
Execute inserts inside ExecutorService

mgazanayi commented:
We can ensure some back pressure by adding a `LinkedBlockingQueue<List<String>>` so we every core doesn't get much more than what it can process.

fbiville commented:
How long does it take to import now? (roughly)

fbiville commented:
With a few fixes/adjustments*, it works great!

Before: 724 seconds
After: 494 seconds

* here is what I had to change:

1. make the batch import call wait via a count down latch until the import is done, else the `:command` would complete in a few seconds and the import would continue in the background
2. keep the import of relationships sequential and with the same batch as before your PR (i.e.1000), Neo4j would otherwise deadlock

Commit:
Fix parallel import

 - concurrentlyImportInBatch should wait for its completion
 - relationships cannot be imported in parallel because it causes
deadlocks

fbiville commented:
PR updated with my fixes. Let me know if you see further improvements @mgazanayi (apart from going with `LOAD CSV` and parameterized queries of course).
